### PR TITLE
[AMDGPU] Add a type for the named barrier

### PR DESCRIPTION
**Pull request by cmc-rep**

### CODE REVIEW

File: clang/include/clang/Basic/AMDGPUTypes.def [hunk 1]
```diff
@@ -15,7 +15,15 @@
...
+  AMDGPU_TYPE(Name, Id, SingletonId, Width, Align)
+#endif
+
 AMDGPU_OPAQUE_PTR_TYPE("__amdgpu_buffer_rsrc_t", AMDGPUBufferRsrc, AMDGPUBufferRsrcTy, 128, 128, 8)
 
+AMDGPU_NAMED_BARRIER_TYPE("__amdgpu_named_workgroup_barrier_t", AMDGPUNamedWorkgroupBarrier, AMDGPUNamedWorkgroupBarrierTy, 128, 32, 0)
+
 #undef AMDGPU_TYPE
 #undef AMDGPU_OPAQUE_PTR_TYPE
+#undef AMDGPU_NAMED_BARRIER_TYPE
```
Comment (jayfoad) [inline clang/include/clang/Basic/AMDGPUTypes.def:29]: Please fix the missing newline.

**Review by jayfoad — APPROVED** [review]
Looks reasonable with the newline fix, but please wait a day in case other reviewers have comments.

**Review by yxsamliu — APPROVED** [review]

**Comment by jplehr** [comment]
Hi,
This broke our OpenMP / Offload bots. It would be great if that can be easily fixed or reverted if more time is required to inspect the issue.
Thanks!

**Comment by jansvoboda11** [comment]
I just pushed a fix in 75252e2.

**Comment by Endilll** [comment]
How did this get through pre-merge testing, I wonder.

**Comment by jansvoboda11** [comment]
This PR did bump `NUM_PREDEF_TYPE_IDS`, but after the pre-merge tests succeeded, #97277 that also bumps `NUM_PREDEF_TYPE_IDS ` landed in `main`.

**Comment by Endilll** [comment]
Ah, so it's the issue of PRs not being tested against `main`.

**Comment by jplehr** [comment]
> I just pushed a fix in [75252e2](https://github.com/llvm/llvm-project/commit/75252e29ea6a0959f3c1670e641a03fc18fc65fa).

I see another error in one of our bots for this patch: https://lab.llvm.org/staging/#/builders/130/builds/7112

**Comment by cmc-rep** [comment]
Looks like I also need to fix the number in that test

**Comment by cmc-rep** [comment]
fix to the test no-external-type-id.cppm: https://github.com/llvm/llvm-project/pull/113738

File: llvm/lib/IR/Type.cpp [hunk 2]
```diff
@@ -839,6 +839,14 @@ Expected<TargetExtType *> TargetExtType::checkParams(TargetExtType *TTy) {
         "target extension type riscv.vector.tuple should have one "
         "type parameter and one integer parameter");
 
+  // Opaque types in the AMDGPU name space.
+  if (TTy->Name == "amdgcn.named.barrier" &&
+      (TTy->getNumTypeParameters() != 0 || TTy->getNumIntParameters() != 1)) {
```
Comment (jayfoad) [inline llvm/lib/IR/Type.cpp:844]: This could also be a good place to check that the "scope" argument is 0, if that is the only supported value.

**Pull request merged**